File: CS/GridControlDetailView/frmMain.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Windows.Forms;
using System.Data;

namespace CustomGridControl
{
    public partial class frmMain : Form
    {
        string[] views = new string[] { "CustomGridControl.MyTreeList", "CustomGridControl.MyVGrid" };

        public frmMain()
        {
            InitializeComponent();
            for (int i = 0; i < views.Length; i++)
                radioGroup1.Properties.Items.Add(new DevExpress.XtraEditors.Controls.RadioGroupItem(views[i], views[i]));
            radioGroup1.Properties.Columns = views.Length;
            radioGroup1.SelectedIndex = 0;
            DataSet set = GetMasterDetail();
            customGridControl1.DataSource = set.Tables["Parent"];
            customView1.InternalControlType = views[radioGroup1.SelectedIndex];
        }

        private DataTable CreateTable(int RowCount)
        {
            DataTable tbl = new DataTable("Parent");
            tbl.Columns.Add("Name", typeof(string));
            tbl.Columns.Add("NOID", typeof(int));
            tbl.Columns.Add("Number", typeof(int));
            tbl.Columns.Add("Date", typeof(DateTime));
            tbl.Columns.Add("DETAILID", typeof(int));
            for (int i = 0; i < RowCount; i++)
                tbl.Rows.Add(new object[] { String.Format("Name{0}", i), i, 3 - i, DateTime.Now.AddDays(i), i });
            return tbl;
        }

        private DataTable CreateDet1Table(int RowCount)
        {
            DataTable tbl = new DataTable("Det1");
            tbl.Columns.Add("Name", typeof(string));
            tbl.Columns.Add("DETAILID", typeof(int));
            tbl.Columns.Add("ID", typeof(int));
            tbl.Columns.Add("ParentID", typeof(int));
            int id = 0;
            for (int j = 0; j < RowCount; j++)
                for (int i = 0; i < RowCount; i++)
                {
                    tbl.Rows.Add(new object[] { String.Format("Detail{1}Name{0}", i, j), i, id, -1 });
                    int parentID = id;
                    id++;
                    CreateSubRows(tbl, RowCount, i, parentID, ref id);
                }
            return tbl;
        }

        private DataTable CreateDet2Table(int RowCount) {
            DataTable tbl = new DataTable("Det2");
            tbl.Columns.Add("Name", typeof(string));
            tbl.Columns.Add("DETAILID", typeof(int));
            tbl.Columns.Add("ID", typeof(int));
            tbl.Columns.Add("ParentID", typeof(int));
            int id = 0;
            for (int j = 0; j < RowCount; j++)
                for (int i = 0; i < RowCount; i++) {
                    tbl.Rows.Add(new object[] { String.Format("Detail{1}Name{0}", i, j), i, id, -1 });
                    int parentID = id;
                    id++;
                    CreateSubRows(tbl, RowCount, i, parentID, ref id);
                }
            return tbl;
        }

        private void CreateSubRows(DataTable tbl, int RowCount, int detailID, int parentID, ref int id)
        {
            for (int i = 0; i < RowCount; i++)
            {
                tbl.Rows.Add(new object[] { String.Format("SubDetail{1}Name{0}", i, parentID), detailID, id++, parentID });
            }
        }

        private DataSet GetMasterDetail()
        {
            DataSet ds = new DataSet("TestDS");
            ds.Tables.Add(CreateTable(20));
            ds.Tables.Add(CreateDet1Table(20));
            ds.Tables.Add(CreateDet2Table(20));
            DataColumn parentColumn = ds.Tables["Parent"].Columns["DETAILID"];
            DataColumn childColumn = ds.Tables["Det1"].Columns["DETAILID"];
            ds.Relations.Add(new DataRelation("relDet1", parentColumn, childColumn));
            parentColumn = ds.Tables["Parent"].Columns["DETAILID"];
            childColumn = ds.Tables["Det2"].Columns["DETAILID"];
            ds.Relations.Add(new DataRelation("relDet2", parentColumn, childColumn));
            return ds;
        }

        private void radioGroup1_SelectedIndexChanged(object sender, EventArgs e)
        {
            customView1.InternalControlType = views[radioGroup1.SelectedIndex];
        }
    }
}

File: VB/GridControlDetailView/frmMain.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Windows.Forms
Imports System.Data

Namespace CustomGridControl
	Partial Public Class frmMain
		Inherits Form
		Private views() As String = { "CustomGridControl.MyTreeList", "CustomGridControl.MyVGrid" }

		Public Sub New()
			InitializeComponent()
			For i As Integer = 0 To views.Length - 1
				radioGroup1.Properties.Items.Add(New DevExpress.XtraEditors.Controls.RadioGroupItem(views(i), views(i)))
			Next i
			radioGroup1.Properties.Columns = views.Length
			radioGroup1.SelectedIndex = 0
			Dim [set] As DataSet = GetMasterDetail()
			customGridControl1.DataSource = [set].Tables("Parent")
			customView1.InternalControlType = views(radioGroup1.SelectedIndex)
		End Sub

		Private Function CreateTable(ByVal RowCount As Integer) As DataTable
			Dim tbl As New DataTable("Parent")
			tbl.Columns.Add("Name", GetType(String))
			tbl.Columns.Add("NOID", GetType(Integer))
			tbl.Columns.Add("Number", GetType(Integer))
			tbl.Columns.Add("Date", GetType(DateTime))
			tbl.Columns.Add("DETAILID", GetType(Integer))
			For i As Integer = 0 To RowCount - 1
				tbl.Rows.Add(New Object() { String.Format("Name{0}", i), i, 3 - i, DateTime.Now.AddDays(i), i })
			Next i
			Return tbl
		End Function

		Private Function CreateDet1Table(ByVal RowCount As Integer) As DataTable
			Dim tbl As New DataTable("Det1")
			tbl.Columns.Add("Name", GetType(String))
			tbl.Columns.Add("DETAILID", GetType(Integer))
			tbl.Columns.Add("ID", GetType(Integer))
			tbl.Columns.Add("ParentID", GetType(Integer))
			Dim id As Integer = 0
			For j As Integer = 0 To RowCount - 1
				For i As Integer = 0 To RowCount - 1
					tbl.Rows.Add(New Object() { String.Format("Detail{1}Name{0}", i, j), i, id, -1 })
					Dim parentID As Integer = id
					id += 1
					CreateSubRows(tbl, RowCount, i, parentID, id)
				Next i
			Next j
			Return tbl
		End Function

		Private Function CreateDet2Table(ByVal RowCount As Integer) As DataTable
			Dim tbl As New DataTable("Det2")
			tbl.Columns.Add("Name", GetType(String))
			tbl.Columns.Add("DETAILID", GetType(Integer))
			tbl.Columns.Add("ID", GetType(Integer))
			tbl.Columns.Add("ParentID", GetType(Integer))
			Dim id As Integer = 0
			For j As Integer = 0 To RowCount - 1
				For i As Integer = 0 To RowCount - 1
					tbl.Rows.Add(New Object() { String.Format("Detail{1}Name{0}", i, j), i, id, -1 })
					Dim parentID As Integer = id
					id += 1
					CreateSubRows(tbl, RowCount, i, parentID, id)
				Next i
			Next j
			Return tbl
		End Function

		Private Sub CreateSubRows(ByVal tbl As DataTable, ByVal RowCount As Integer, ByVal detailID As Integer, ByVal parentID As Integer, ByRef id As Integer)
			For i As Integer = 0 To RowCount - 1
				tbl.Rows.Add(New Object() { String.Format("SubDetail{1}Name{0}", i, parentID), detailID, id, parentID })
				id += 1
			Next i
		End Sub

		Private Function GetMasterDetail() As DataSet
			Dim ds As New DataSet("TestDS")
			ds.Tables.Add(CreateTable(20))
			ds.Tables.Add(CreateDet1Table(20))
			ds.Tables.Add(CreateDet2Table(20))
			Dim parentColumn As DataColumn = ds.Tables("Parent").Columns("DETAILID")
			Dim childColumn As DataColumn = ds.Tables("Det1").Columns("DETAILID")
			ds.Relations.Add(New DataRelation("relDet1", parentColumn, childColumn))
			parentColumn = ds.Tables("Parent").Columns("DETAILID")
			childColumn = ds.Tables("Det2").Columns("DETAILID")
			ds.Relations.Add(New DataRelation("relDet2", parentColumn, childColumn))
			Return ds
		End Function

		Private Sub radioGroup1_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles radioGroup1.SelectedIndexChanged
			customView1.InternalControlType = views(radioGroup1.SelectedIndex)
		End Sub
	End Class
End Namespace
